src/net/atlas/tests.rs

```diff
@@ -0,0 +1,634 @@
+        AttachmentsBatchStateContext::new(attachments_batch, peers, &ConnectionOptions::default());
+
+    let mut inventories_requests = context.get_prioritized_attachments_inventory_requests();
+    let mut inventories_results = BatchedRequestsResult::empty();
+
+    let request = inventories_requests.pop().unwrap();
+    let peer_url_1 = request.get_url().clone();
+    let request = inventories_requests.pop().unwrap();
+    let peer_url_2 = request.get_url().clone();
+    let request = inventories_requests.pop().unwrap();
+    let peer_url_3 = request.get_url().clone();
+    let request = inventories_requests.pop().unwrap();
+    let peer_url_4 = request.get_url().clone();
+    let mut responses = HashMap::new();
+
+    let response_1 =
+        new_attachments_inventory_response(vec![(1, vec![1, 1, 1]), (2, vec![0, 0, 0])]);
+    responses.insert(peer_url_1.clone(), Some(response_1));
+
+    let response_2 =
+        new_attachments_inventory_response(vec![(1, vec![1, 1, 1]), (2, vec![0, 0, 0])]);
+    responses.insert(peer_url_2.clone(), Some(response_2));
+
+    let response_3 =
+        new_attachments_inventory_response(vec![(1, vec![0, 1, 1]), (2, vec![1, 0, 0])]);
+    responses.insert(peer_url_3.clone(), Some(response_3));
+    responses.insert(peer_url_4, None);
+
+    inventories_results.succeeded.insert(request, responses);
+
+    let context = context.extend_with_inventories(&mut inventories_results);
+
+    let mut attachments_requests = context.get_prioritized_attachments_requests();
+
+    let request = attachments_requests.pop().unwrap();
+    let request_type = request.get_request_type(localhost.clone());
+    // Attachment 4 is the rarest resource
+    assert_eq!(
+        request_type.request_path(),
+        format!("/v2/attachments/{}", attachment_4.hash)
+    );
+    // Peer 1 is the only peer showing Attachment 4 as being available in its inventory
+    assert_eq!(request.get_url(), &peer_url_3);
+
+    let request = attachments_requests.pop().unwrap();
+    let request_type = request.get_request_type(localhost.clone());
+    // Attachment 1 is the 2nd rarest resource
+    assert_eq!(
+        request_type.request_path(),
+        format!("/v2/attachments/{}", attachment_1.hash)
+    );
+    // Both Peer 1 and Peer 2 could serve Attachment 1, but Peer 1 has a better history
+    assert_eq!(request.get_url(), &peer_url_1);
+
+    // The 2 last requests can be served by Peer 1, 2 and 3, but will be served in random
+    // order by Peer 1 (best score).
+    let request = attachments_requests.pop().unwrap();
+    let request_type = request.get_request_type(localhost.clone());
+    assert_eq!(request.get_url(), &peer_url_1);
+
+    let request = attachments_requests.pop().unwrap();
+    let request_type = request.get_request_type(localhost.clone());
+    assert_eq!(request.get_url(), &peer_url_1);
+}
+
+#[test]
+fn test_downloader_dns_state_machine() {}
+
+#[test]
+fn test_downloader_batched_requests_state_machine() {}
```

Comment:
There needs to be some tests to cover the following classes of scenarios:

* Names get registered, updated, etc. in BNS, generating a history of zone file hashes.  The code should verify that it can extract an accurate zone file inventory bit vector, given which zone file hashes have an associated attachment.
* A node can download attachments from another node as long as it shares the same zone file inventory vector and chain tip.
* Nodes that do not agree on the same chain tip should not exchange zone files, or perhaps only exchange zone files for the part of the chain history they do agree on.
* A node that bootstraps can incrementally download zone files for the portion of the chain history it knows about.

Reply:
I see that you've added tests for name registration, updates, etc.  Thanks!